shared/prow/prow.go

```diff
@@ -0,0 +1,336 @@
+	if dir == "" {
+		log.Printf("Env variable ARTIFACTS not set. Using %s instead.", ArtifactsDir)
+		dir = ArtifactsDir
+	}
+	return dir
+}
+
+
+/* GCS related logics */
+
+// Initialize wraps gcs authentication, have to be invoked before any other functions
+func Initialize(sa string) error {
+	return gcs.Authenticate(ctx, sa)
+}
+
+// NewJob creates new job struct
+func NewJob(jobName string) *Job {
+	return &Job{
+		Name: jobName,
+		Bucket: BucketName,
+	}
+}
+
+// NewBuild creates new build struct
+func NewBuild(jobName string, buildID int) *Build {
+	return &Build{
+		Bucket: BucketName,
+		JobName: jobName,
+		BuildID: buildID,
+	}
+}
+
+// GetPath gets gcs path of current job
+func (j *Job) GetPath() string {
+	if nil != j.Prefix { // return custom path if set
+		return *j.Prefix
+	}
+
+	var res string
+	switch j.Type {
+		case PeriodicJob, PostsubmitJob:
+			res = path.Join("logs", j.Name)
+		case PresubmitJob:
+			res = path.Join("pr-logs", "pull", OrgName + "-" + j.Repo, j.Commit, j.Name)
+		case BatchJob:
+			res = path.Join("pr-logs", "pull", "batch", j.Name)
+		default:
+			log.Fatalf("unknown job spec type: %v", j.Type)
+	}
+
+	// cache prefix if computed
+	j.Prefix = &res
+	return res
+}
+
+// GetLatestBuildNumber gets the latest build number for job
+func (j *Job) GetLatestBuildNumber() (int, error) {
+	logFilePath := path.Join(j.GetPath(), Latest)
+	log.Printf("Using %s to get latest build number", logFilePath)
+	contents, err := gcs.Read(ctx, BucketName, logFilePath)
+	if err != nil {
+		return 0, err
+	}
+	latestBuild, err := strconv.Atoi(strings.TrimSuffix(string(contents), "\n"))
+	if err != nil {
+		return 0, err
+	}
+
+	return latestBuild, nil
+}
+
+// GetBuild gets build struct based on job info
+// 	No gcs operation
+func (j *Job) GetBuild(buildID int) *Build {
+	return &Build{
+		Bucket: BucketName,
+		JobName: j.Name,
+		JobType: j.Type,
+		BuildID: buildID,
+	}
+}
+
+// GetFinishedBuilds gets all builds that have finished,
+// by looking at existence of "finished.json" file
+func (j *Job) GetFinishedBuilds() []Build {
+	var finishedBuilds []Build
+	builds := j.GetBuilds()
+	for _, build := range builds {
+		if true == build.IsFinished() {
+			finishedBuilds = append(finishedBuilds, build)
+		}
+	}
+	return finishedBuilds
+}
+
+// GetBuilds gets all builds from this job on gcs
+func (j *Job) GetBuilds() []Build {
+	var builds []Build
+	gcsBuildPaths := gcs.ListDirectChildren(ctx, j.Bucket, j.GetPath())
+	for _, gcsBuildPath := range gcsBuildPaths {
+		buildID, err := GetBuildIDFromBuildPath(gcsBuildPath)
+		if nil != err { // this last part of gcs path is not a valid int64, should not be a build
+			continue
+		}
+		build := Build{
+			Bucket: j.Bucket,
+			JobName: j.Name,
+			BuildID: buildID,
+		}
+		if true == build.IsStarted() {
+			builds = append(builds, build)
+		}
+	}
+	return builds
+}
+
+// GetLatestBuilds get latest builds from gcs
+// The timestamp of gcs directories are not usable, 
```

Comment:
This comment should be inside the function, not here.

Reply:
done